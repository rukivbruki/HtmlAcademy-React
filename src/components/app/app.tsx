// Libs
import * as React from 'react';
import {connect} from "react-redux";
import {Route, Switch, withRouter} from "react-router-dom";
// Utils
import Paths from '../../router/paths';
import {FilmsType, FilmType, HistoryType} from "../../types";
import {getFilmsByGenre} from "../../reducer/films/selectors";
import {getGenreFilter} from "../../reducer/genres/selectors";
import {getFilms} from "../../reducer/films/selectors";
import {getUser, getIsAuth} from "../../reducer/user/selectors.js";
import withProgress from "../../hocs/with-progress/with-progress";
import withInputs from "../../hocs/with-inputs/with-inputs";
import withCheckAuth from "../../hocs/with-check-auth/with-check-auth";
import withReviewData from "../../hocs/with-review-data/with-review-data";
// Components
import Main from "../main/main";
import MovieDetails from "../movie-details/movie-details";
import VideoPlayer from "../video-player/video-player";
import SignIn from "../sign-in/sign-in";
import AddReview from "../add-review/add-review";
import MyList from "../my-list/my-list";
import PrivateRoute from '../private-route/private-route';
import {getPromo} from "../../reducer/promo/selectors";

const VideoPlayerWithProgress = withProgress(VideoPlayer);
const SignInWithInputs = withCheckAuth(withInputs(SignIn));
const RewiewWithReviewData = withReviewData(AddReview);

interface Props {
  setActiveItem: (data: FilmType) => void;
  setActivePlayer: (activePlayer: boolean) => void;
  filteredFilms: FilmsType;
  activeItem?: FilmType;
  promo: FilmType;
  history?: HistoryType;
  films: FilmsType;
  activePlayer?: boolean;
}

class App extends React.PureComponent<Props> {
  constructor(props) {
    super(props);

    this._handleOpenCard = this._handleOpenCard.bind(this);
    this._renderVideoPlayer = this._renderVideoPlayer.bind(this);
    this._renderApp = this._renderApp.bind(this);
  }

  _handleOpenCard(data: FilmType) {
    const {setActiveItem} = this.props;
    setActiveItem(data);
  }

  _renderVideoPlayer() {
    const {setActivePlayer, films} = this.props;
    return <VideoPlayerWithProgress setActivePlayer={setActivePlayer} films={films}/>;
  }

  _renderApp() {
    const {filteredFilms, activeItem, promo, setActivePlayer, history, films} = this.props;
    return (
      <Switch>
        <Route exact path={Paths.INDEX}>
          <Main
            {...this.props}
            onOpenCard={this._handleOpenCard}
            filteredFilms={filteredFilms}
            promoData={promo}
            history={history}
          />
        </Route>
        <Route path={Paths.FILMS_$ID_PLAYER}>
          <VideoPlayerWithProgress setActivePlayer={setActivePlayer} history={history} films={films}/>
        </Route>
        <PrivateRoute exact path={Paths.FILMS_$ID_REVIEW} render={() => (
          <RewiewWithReviewData />
        )}>
        </PrivateRoute>
        <Route path={Paths.FILMS_$ID}>
          <MovieDetails
            {...this.props}
            cardData={activeItem}
            onOpenCard={this._handleOpenCard}
            filteredFilms={filteredFilms}
            setActivePlayer={setActivePlayer}
          />
        </Route>
        <Route path={Paths.LOGIN}>
          <SignInWithInputs />
        </Route>
        <PrivateRoute exact path={Paths.MY_LIST} render={() => (
          <MyList />
        )}>
        </PrivateRoute>
      </Switch>
    );
  }

  render() {
    return this.props.activePlayer
      ? this._renderVideoPlayer()
      : this._renderApp();
  }
}

const mapStateToProps = (state) => ({
  films: getFilms(state),
  genreFilter: getGenreFilter(state),
  filteredFilms: getFilmsByGenre(state),
  user: getUser(state),
  isAuth: getIsAuth(state),
  promo: getPromo(state)
});

const AppWrapper = connect(mapStateToProps)(App);

export {App};
export default withRouter(AppWrapper);
