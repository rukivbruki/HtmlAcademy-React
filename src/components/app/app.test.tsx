import * as React from "react";
import * as renderer from "react-test-renderer";
import {Provider} from "react-redux";
import {BrowserRouter} from "react-router-dom";
import configureStore from "redux-mock-store";
import NameSpace from "../../reducer/name-space.js";
import {App} from './app';

const mockStore = configureStore([]);

const noop = () => void 0;

const film = {
  backgroundColor: `red`,
  backgroundImage: `img`,
  description: `desc`,
  director: `dir`,
  genre: `genre`,
  id: 1,
  isFavorite: false,
  name: `film`,
  posterImage: `img`,
  previewImage: `img`,
  previewVideoLink: `video`,
  rating: 1,
  released: 1,
  runTime: 1,
  scoresCount: 1,
  starring: [`test`],
  videoLink: `video`
};

const films = [film];

it(`App renders correctly`, () => {
  const store = mockStore({
    [NameSpace.GENRES]: {
      genres: [`genre1`, `genre2`],
      genreFilter: `All genres`
    },
    [NameSpace.FILMS]: {
      films
    },
    [NameSpace.PROMO]: {
      promo: films[0]
    },
    [NameSpace.USER]: {
      authorizationStatus: `AUTH`,
      user: {
        id: 1,
        email: `test@email.com`,
        name: `test user`,
        avatarUrl: `avatar`
      }
    },
    [NameSpace.FAVORITES]: {
      favorites: films
    },
  });

  const props = {
    setActiveItem: noop,
    setActivePlayer: noop,
    filteredFilms: [film],
    promo: film,
    films,
  };

  const tree = renderer
    .create(
        <Provider store={store}>
          <BrowserRouter>
            <App {...props}/>
          </BrowserRouter>
        </Provider>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
