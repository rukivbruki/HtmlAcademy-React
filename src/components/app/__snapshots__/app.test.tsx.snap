// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
Array [
  <section
    className="movie-card"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="film"
        src="img"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="user-block"
      >
        <div
          className="user-block__avatar"
          onClick={[Function]}
        >
          <img
            alt="User avatar"
            height="63"
            src="https://htmlacademy-react-3.appspot.comavatar"
            width="63"
          />
        </div>
      </div>
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster"
        >
          <img
            alt="The Grand Budapest Hotel poster"
            height="327"
            src="img"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            film
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              genre
            </span>
            <span
              className="movie-card__year"
            >
              1
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                fill="#d9cd8d"
                height="24"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 5v14l11-7z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="14"
                onClick={[Function]}
                viewBox="0 0 18 14"
                width="18"
              >
                <use
                  xlinkHref="#in-list"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        className="catalog__genres-list"
      >
        <li
          className="catalog__genres-item "
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            genre1
          </a>
        </li>
        <li
          className="catalog__genres-item "
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            genre2
          </a>
        </li>
      </ul>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "position": "relative",
            }
          }
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="film"
              height="175"
              src="img"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
            style={
              Object {
                "height": "100%",
                "width": "100%",
              }
            }
          >
            <a
              className="small-movie-card__link"
              href="/films/1"
              onClick={[Function]}
              style={
                Object {
                  "alignItems": "flex-end",
                  "display": "flex",
                  "height": "100%",
                  "width": "100%",
                }
              }
            >
              film
            </a>
          </h3>
        </article>
      </div>
      <div
        className="catalog__more"
      />
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
          href="/"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
